README.md


## Getting started

To use this plugin, add checkbox_list as a dependency in your pubspec.yaml file.

This library currently has one Widget.

*CustomCheckBox

## Usage
```rb
import 'package:custom_checkbox/custom_checkbox.dart';
import 'package:flutter/material.dart';


void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom checkbox Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const CheckBoxDemo(),
    );
  }
}

class CheckBoxDemo extends StatelessWidget {
  const CheckBoxDemo({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoise: false,
              textLabel: "Value 1",
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.blue,
              onChanged: (value) {
                print(value);
              },
              clickableLabelSplashColor: Colors.blue, //In order to use this property, set clickable label to true
              clickableLabel: true,
              multipleChoise: false,
              textLabel: "Value 2",
              labelTextStyle:
              const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoise: false,
              textLabel: "Value 3",
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
              checkBoxSplashColor: Colors.blue,
              checkBoxSplashRadius: 20,
            ),

            //single choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoise: false,
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //multiple choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoise: true,
              clickableLabel: true, //for clicking the whole container
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //single choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.green,
              onChanged: (value) {
                print(value);
              },
              shape: BoxShape.circle,
              checkBoxSplashRadius: 15,
              checkBoxSplashColor: Colors.yellow,
              multipleChoise: false,
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
              selectedBoxColor: Colors.red,
            ),
          ],
        ),
      ),
    );
  }
}
```




CHANGELOG.md
## 0.0.17

*  New property for margin between the checkbox list children


lib/src/custom_check_box.dart
import 'package:flutter/material.dart';

class CustomCheckBox extends StatefulWidget {
  final List<String>? checkboxList;
  final Color borderColor;
  final Color checkIconColor;
  final Color selectedBoxColor;
  final TextStyle? labelTextStyle;
  final double paddingBetweenTitleAndCheckBox;
  final Widget? widgetUnderTheLabel;
  final double checkBoxWidth;
  final double checkBoxHeight;
  final bool multipleChoice;
  final Function(dynamic) onChanged;
  final Icon? checkBoxIcon;
  final double iconSize;
  final double checkBoxSplashRadius;
  final bool clickableLabel;
  final double borderWidth;
  final Color? checkBoxSplashColor;
  final Color? clickableLabelSplashColor;
  final String? textLabel;
  final BorderRadius? borderRadius;
  final BoxShape shape;
  final EdgeInsets margin;

  CustomCheckBox({
    super.key,
    this.checkboxList,
    this.borderColor = Colors.black,
    this.labelTextStyle,
    this.checkIconColor = Colors.black,
    this.selectedBoxColor = Colors.transparent,
    this.paddingBetweenTitleAndCheckBox = 0,
    this.widgetUnderTheLabel,
    this.checkBoxWidth = 15,
    this.checkBoxHeight = 15,
    required this.onChanged,
    required this.multipleChoice,
    this.checkBoxIcon,
    this.iconSize = 1.4,
    this.checkBoxSplashRadius = 10,
    this.clickableLabel = false,
    this.borderWidth = 2.0,
    this.checkBoxSplashColor,
    this.clickableLabelSplashColor,
    this.textLabel,
    this.borderRadius,
    this.shape = BoxShape.rectangle,
    this.margin = EdgeInsets.zero,
  });

  @override
  State<CustomCheckBox> createState() => _CustomCheckBoxState();
}

class _CustomCheckBoxState extends State<CustomCheckBox> {
  String selected = "";
  List<Map<String, dynamic>> list = [];
  List<String> selectedList = [];
  bool isChecked = false;

  @override
  void initState() {
    super.initState();
    if (widget.checkboxList != null) {
      for (var i = 0; i < widget.checkboxList!.length; i++) {
        list.add({"id": i, "bool": false, "value": widget.checkboxList![i]});
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return widget.checkboxList == null && widget.textLabel == null
        ? InkResponse(
            radius: widget.checkBoxSplashRadius,
            //in order to use this property set the clickableLabel property to false
            splashColor: widget.checkBoxSplashColor,
            onTap: !widget.clickableLabel
                ? () {
                    onTap(0);
                  }
                : null,
            child: UnconstrainedBox(
              child: Container(
                width: widget.checkBoxWidth,
                height: widget.checkBoxHeight,
                alignment: Alignment.center,
                decoration: BoxDecoration(
                    color: isChecked == true ? widget.selectedBoxColor : null,
                    border: Border.all(
                      color: widget.borderColor,
                      width: widget.borderWidth,
                    ),
                    borderRadius: widget.borderRadius,
                    shape: widget.borderRadius == null
                        ? widget.shape
                        : BoxShape.rectangle),
                child: isChecked == true
                    ? FittedBox(
                        child: Transform.scale(
                          scale: widget.iconSize,
                          child: widget.checkBoxIcon ??
                              Icon(
                                Icons.check,
                                color: widget.checkIconColor,
                              ),
                        ),
                      )
                    : null,
              ),
            ),
          )
        : widget.checkboxList == null && widget.textLabel != null
            ? InkWell(
                splashColor: widget.clickableLabelSplashColor,
                //in order to use this property set the clickableLabel property to true
                onTap: widget.clickableLabel
                    ? () {
                        onTap(0);
                      }
                    : null,
                child: Row(
                  children: [
                    InkResponse(
                      radius: widget.checkBoxSplashRadius,
                      //in order to use this property set the clickableLabel property to false
                      splashColor: widget.checkBoxSplashColor,
                      onTap: !widget.clickableLabel
                          ? () {
                              onTap(0);
                            }
                          : null,
                      child: Container(
                        width: widget.checkBoxWidth,
                        height: widget.checkBoxHeight,
                        alignment: Alignment.center,
                        decoration: BoxDecoration(
                            color: isChecked == true
                                ? widget.selectedBoxColor
                                : null,
                            border: Border.all(
                              color: widget.borderColor,
                              width: widget.borderWidth,
                            ),
                            borderRadius: widget.borderRadius,
                            shape: widget.borderRadius == null
                                ? widget.shape
                                : BoxShape.rectangle),
                        child: isChecked == true
                            ? FittedBox(
                                child: Transform.scale(
                                  scale: widget.iconSize,
                                  child: widget.checkBoxIcon ??
                                      Icon(
                                        Icons.check,
                                        color: widget.checkIconColor,
                                      ),
                                ),
                              )
                            : null,
                      ),
                    ),
                    Padding(
                      padding: EdgeInsets.only(
                          left: widget.paddingBetweenTitleAndCheckBox),
                      child: Wrap(
                          direction: Axis.vertical,
                          spacing: widget.widgetUnderTheLabel == null ? 0 : -5,
                          children: [
                            Text(widget.textLabel!,
                                style:
                                    widget.labelTextStyle ?? const TextStyle()),
                            widget.widgetUnderTheLabel ?? const SizedBox(),
                          ]),
                    ),
                  ],
                ),
              )
            : Column(
                children: List.generate(
                    list.length,
                    (index) => Padding(
                          padding: widget.margin,
                          child: InkWell(
                            splashColor: widget.clickableLabelSplashColor,
                            //in order to use this property set the clickableLabel property to true
                            onTap: widget.clickableLabel
                                ? () {
                                    onTap(index);
                                  }
                                : null,
                            splashFactory:
                                InkSparkle.constantTurbulenceSeedSplashFactory,
                            child: Row(
                              children: [
                                InkResponse(
                                  radius: widget.checkBoxSplashRadius,
                                  //in order to use this property set the clickableLabel property to false
                                  splashColor: widget.checkBoxSplashColor,
                                  onTap: !widget.clickableLabel
                                      ? () {
                                          onTap(index);
                                        }
                                      : null,
                                  child: Container(
                                    width: widget.checkBoxWidth,
                                    height: widget.checkBoxHeight,
                                    alignment: Alignment.center,
                                    decoration: BoxDecoration(
                                        color: list[index]["bool"] == true
                                            ? widget.selectedBoxColor
                                            : null,
                                        border: Border.all(
                                          color: widget.borderColor,
                                          width: widget.borderWidth,
                                        ),
                                        borderRadius: widget.borderRadius,
                                        shape: widget.borderRadius == null
                                            ? widget.shape
                                            : BoxShape.rectangle),
                                    child: list[index]["bool"] == true
                                        ? FittedBox(
                                            child: Transform.scale(
                                              scale: widget.iconSize,
                                              child: widget.checkBoxIcon ??
                                                  Icon(
                                                    Icons.check,
                                                    color:
                                                        widget.checkIconColor,
                                                  ),
                                            ),
                                          )
                                        : null,
                                  ),
                                ),
                                Padding(
                                  padding: EdgeInsets.only(
                                      left: widget
                                          .paddingBetweenTitleAndCheckBox),
                                  child: Wrap(
                                      direction: Axis.vertical,
                                      spacing:
                                          widget.widgetUnderTheLabel == null
                                              ? 0
                                              : -5,
                                      children: [
                                        Text(list[index]["value"],
                                            style: widget.labelTextStyle ??
                                                const TextStyle()),
                                        widget.widgetUnderTheLabel ??
                                            const SizedBox()
                                      ]),
                                ),
                              ],
                            ),
                          ),
                        )),
              );
  }

  onTap(int index) {
    if (widget.checkboxList != null) {
      setState(() {
        selected = list[index]["value"];
        list[index]["bool"] = !list[index]["bool"];

        if (!widget.multipleChoice) {
          for (var i = 0; i < list.length; i++) {
            if (list[i]['bool'] == true && selected != list[i]['value']) {
              list[i]['bool'] = false;
              i = list.length;
            }
          }
          //single string value(comment this if you want a value with list of string)
          if (list[index]["bool"] == false) selected = "";

          //you can change the argument type of this function if you want to send a different value
          widget.onChanged(selected);
        } else {
          if (list[index]["bool"] == true) selectedList.add(selected);
          if (list[index]["bool"] == false) selectedList.remove(selected);
          //if multiple selection is true list of string will be setup as a value
          widget.onChanged(selectedList);
        }
      });
    } else {
      //if the checkbox list is empty a boolean will be set as an argument to onChanged function
      setState(() {
        isChecked = !isChecked;
        widget.onChanged(isChecked);
      });
    }
  }
}


example/main.dart
import 'package:checkbox_list/checkbox_list.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom checkbox Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const CheckBoxDemo(),
    );
  }
}

class CheckBoxDemo extends StatelessWidget {
  const CheckBoxDemo({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoice: false,
              textLabel: "Value 1",
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.blue,
              onChanged: (value) {
                print(value);
              },
              clickableLabelSplashColor: Colors.blue, //In order to use this property, set clickable label to true
              clickableLabel: true,
              multipleChoice: false,
              textLabel: "Value 2",
              labelTextStyle:
              const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //labeled single choice checkbox
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoice: false,
              textLabel: "Value 3",
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
              checkBoxSplashColor: Colors.blue,
              checkBoxSplashRadius: 20,
            ),

            //single choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoice: false,
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //multiple choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.orange,
              onChanged: (value) {
                print(value);
              },
              multipleChoice: true,
              clickableLabel: true, //for clicking the whole container
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
            ),
            //single choice with check box list
            CustomCheckBox(
              checkBoxWidth: 20,
              checkBoxHeight: 20,
              borderColor: Colors.green,
              onChanged: (value) {
                print(value);
              },
              shape: BoxShape.circle,
              checkBoxSplashRadius: 15,
              checkBoxSplashColor: Colors.yellow,
              multipleChoice: false,
              checkboxList: [
                "1",
                "2",
                "3",
                "4"
              ], //setup a list of strings to generate a list of labeled checkboxes
              labelTextStyle:
                  const TextStyle(color: Colors.black, fontSize: 16),
              paddingBetweenTitleAndCheckBox: 13,
              selectedBoxColor: Colors.red,
            ),
          ],
        ),
      ),
    );
  }
}
